wurst/network/Network.wurst

```diff
@@ -0,0 +1,922 @@
+@configurable constant integer MAX_CHARACTERS_PER_KEY = 3
+// setting this to false will optimize out all safety checks...
+// only disable this if you are 100% sure your code works correctly
+@configurable constant boolean SAFETY_CHECKS_ENABLED = true
+
+//TODO: I just realized that we can eliminate some delay by sending metadata and the first round together
+// and avoid one round-trip, essentially reducing transmission time for small payloads up to two times
+
+/********************************************************************************
+	Network library v1.0 by MoriMori.
+
+	Based on TriggerHappy's Sync library.
+
+	Rationale:
+		This library can be used to send arbitrary amounts of integer, real, boolean
+		and string data from one player to the rest.
+
+		Like SyncSimple, it abuses the fact that all network actions in WC3 are
+		sequential, and are received by players in the same order that are sent in.
+
+		It uses the SyncStored* natives of the gamecache to send data to other players
+		from the sender, and then uses SyncSimple to send a 'finish' event, that will
+		only be received by other players after they have also received all sync data.
+
+		The Network class provides 4 independent streams of data:
+			integers,
+			reals,
+			booleans,
+			strings
+
+		each of which can be written/read to using the relevant 
+		methods of the DataStream class.
+
+		Before sending the data, the sender populates the DataStream with the data that
+		they want to send, then Network.start() is called, as well as Network.onFinish(callback)
+		and data is received from the same buffer inside the callback when it has all been received.
+
+		Read SyncSimple docs for a slightly more in-depth overview of this particular
+		peculiarity of WC3.
+
+	Usage:
+		1. Create a new Network object using:
+			// sender is the player to receive data from
+			let network = new Network(sender)
+
+		2. Queue some data for sending from the sender. You can write as much as you need, there is no limit:
+			if localPlayer == sender
+				// keep in mind that before you start sending data,
+				// network.getData() is locked in a WRITE-ONLY MODE
+				val stream = network.getData()
+				stream.writeInteger(...)
+				stream.writeReal(...)
+				stream.writeString(...)
+				...
+
+		3. To start sending the data, simply call:
+			network.start()			
+			// this will start the transmission
+			// keep in mind, that after you call this, you MUST NOT write/read to network.getData()
+			// as it is locked while the data is syncing
+		
+		4. To receive the data, specify the callback and read the data inside it:
+			network.onFinish((stream) -> begin
+				// here, network.getData() is locked in a READ-ONLY MODE and
+				// you can only READ from it
+				var int = stream.readInteger()
+				var real = stream.readReal()
+				var string = stream.readString()
+				...
+			end)
+
+			WARNING: Network is a one-time use class, and is automatically
```

Comment:
You mean: `start()` performs a `destroy this`?

Reply:
`destroy this` is performed after the callback specified in `.onFinish(cb)` is called.

Reply:
I'm wondering if this warning is necessary. Will wurst compiler prevent you from use-after-free? @Frotty